lib/components/sync-status/get-note-titles.ts

```diff
@@ -1,24 +1,25 @@
-import { compact } from 'lodash';
 import noteTitleAndPreview from '../../utils/note-utils';
 
-const getNoteTitles = (ids, notes, limit = Infinity) => {
-  const matchedNotes = ids.map((id, i) => {
-    if (i >= limit) {
-      return;
-    }
+import * as T from '../../types';
 
-    const note = notes.find(thisNote => thisNote.id === id);
-
-    if (!note) {
-      // eslint-disable-next-line no-console
-      console.log(`Could not find note with id '${id}'`);
-      return null;
-    }
+type NoteTitle = {
+  id: String;
+  title: String;
+};
 
-    return { id, title: noteTitleAndPreview(note).title };
-  });
+const getNoteTitles = (
+  ids: String[] = [],
+  notes: T.NoteEntity[] = [],
+  limit: number = Infinity
+) => {
+  const matchedNotes = ids.reduce((acc: NoteTitle[], id: String) => {
+    const note = notes.find((thisNote: T.NoteEntity) => thisNote.id === id);
+    return note
+      ? [...acc, { id, title: noteTitleAndPreview(note).title }]
+      : acc;
+  }, []);
 
-  return compact(matchedNotes);
+  return matchedNotes.slice(0, limit);
```

Comment:
It looks like we're doing two things in this PR: (a) fixing a bug due to not checking for `null`; and (b) refactoring the `getNoteTitles` function.

If we want to focus on the bug it might be easier to do that first and then come back for a refactor so that we're not fixing a defect _and_ changing behaviors in the same patch (as we're doing with the move from `null` + `compact()` to the `reducer()`)

If we want to go ahead and refactor we might consider a variation of the function that doesn't involve the `reduce()`. Inside that reducer we're iterating across notes and filtering and mapping all at once, and it will get progressively slower as we get more notes because of the loop-inside-a-loop. We could separate the duties and do one quick setup to remove the inner loop. I'm borrowing `filterAtMost()` from `tag-suggestions/index` but it's comparable to `.filter().slice()` and just skips processing any more after we get our limit.

```ts
const getNoteTitles = (
  ids: T.EntityId[] = [],
  notes: T.NoteEntity[] = [],
  limit: number = Infinity
): NoteTitle[] => {
  const wantedIds = new Set(ids);
  const wantedNotes = filterAtMost(notes, ({ id }) => wantedIds.has(id), limit);

  return wantedNotes.map(note => ({
    id: note.id,
    title: noteTitleAndPreview(note).title,
  }));
};
```

Creating the `Set()` turns our list-scan into a single lookup per id and should only incur the cost of the first iteration through the `notes` array. If we have two or more `notes` then the cost of doing that should be offset (if we only have one then it doesn't really matter anyway).

`filterAtMost()` saves on an intermediate data structure and only iterates across ten `notes` if that's the limit even if we have 10,000 notes in the account.

Reply:
Created another PR for Fix here: https://github.com/Automattic/simplenote-electron/pull/1898

Reply:
Changed in: febd61e

I very much dislike the added complexity of filterAtMost but agreed that it would be more performance.